jxl/src/render/stages/upsample.rs

```diff
@@ -57,6 +80,410 @@ impl<const N: usize, const SHIFT: u8> std::fmt::Display for Upsample<N, SHIFT> {
+
+        // Pre-broadcast kernel weights
+        // flat_kernels layout: kernel[oy][ox] -> flat_kernels[oy * 2 + ox]
+        let mut kernel_vecs = [[D::F32Vec::splat(d, 0.0); 25]; 4];
+        for idx in 0..4 {
+            let k = &flat_kernels[idx];
+            for i in 0..25 {
+                kernel_vecs[idx][i] = D::F32Vec::splat(d, k[i]);
+            }
+        }
+
+        // Process using iterators for mins/maxs, manual indexing for output
+        let mins_iter = mins.chunks_exact(D::F32Vec::LEN);
+        let maxs_iter = maxs.chunks_exact(D::F32Vec::LEN);
+
+        for ((mins_chunk, maxs_chunk), x) in mins_iter
+            .zip(maxs_iter)
+            .zip((0..xsize).step_by(D::F32Vec::LEN))
+            .take(xsize.div_ceil(D::F32Vec::LEN))
+        {
+            let minval = D::F32Vec::load(d, mins_chunk);
+            let maxval = D::F32Vec::load(d, maxs_chunk);
+            let out_x = x * 2;
+
+            // Row 0
+            let r0_0 = kernel_conv!(d, kernel_vecs[0], r0, r1, r2, r3, r4, x).max(minval).min(maxval);
+            let r0_1 = kernel_conv!(d, kernel_vecs[1], r0, r1, r2, r3, r4, x).max(minval).min(maxval);
+            D::F32Vec::store_interleaved_2(r0_0, r0_1, &mut output[0][out_x..]);
+
+            // Row 1
+            let r1_0 = kernel_conv!(d, kernel_vecs[2], r0, r1, r2, r3, r4, x).max(minval).min(maxval);
+            let r1_1 = kernel_conv!(d, kernel_vecs[3], r0, r1, r2, r3, r4, x).max(minval).min(maxval);
+            D::F32Vec::store_interleaved_2(r1_0, r1_1, &mut output[1][out_x..]);
+        }
+    }
+);
+
+// 4x upsampling SIMD implementation
+simd_function!(
+    upsample_4x_simd_dispatch,
+    d: D,
+    fn upsample_4x_simd(
+        input: &[&[f32]],
+        xsize: usize,
+        flat_kernels: &[[f32; 25]],
+        mins: &[f32],
+        maxs: &[f32],
+        output: &mut [&mut [f32]],
+    ) {
+        let r0 = input[0];
+        let r1 = input[1];
+        let r2 = input[2];
+        let r3 = input[3];
+        let r4 = input[4];
+
+        // Pre-broadcast kernel weights
+        // flat_kernels layout: kernel[oy][ox] -> flat_kernels[oy * 4 + ox]
+        let mut kernel_vecs = [[D::F32Vec::splat(d, 0.0); 25]; 16];
+        for idx in 0..16 {
+            let k = &flat_kernels[idx];
+            for i in 0..25 {
+                kernel_vecs[idx][i] = D::F32Vec::splat(d, k[i]);
+            }
+        }
+
+        // Process using iterators for mins/maxs, manual indexing for output
+        let mins_iter = mins.chunks_exact(D::F32Vec::LEN);
+        let maxs_iter = maxs.chunks_exact(D::F32Vec::LEN);
+
+        for ((mins_chunk, maxs_chunk), x) in mins_iter
+            .zip(maxs_iter)
+            .zip((0..xsize).step_by(D::F32Vec::LEN))
+            .take(xsize.div_ceil(D::F32Vec::LEN))
+        {
+            let minval = D::F32Vec::load(d, mins_chunk);
+            let maxval = D::F32Vec::load(d, maxs_chunk);
+            let out_x = x * 4;
+
+            // Row 0
```

Comment:
This starts to be a lot of repeated code - maybe it's best do make a for loop for the rows (also for 8x -- 2x is fine as is).